planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+* Can we consider ditching < .Net Core or .Net v5 for v4? 
+
+
+## Dynamic Code Generation
+
+If you look at the [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364) and the code in `Marten.Schema.Arguments` you can see that our dynamic `Expression` to `Lambda` compilation code is getting extremely messy, hard to reason with, and difficult to extend.
+
+**Idea**: Introduce a dependency on [LamarCodeGeneration and LamarCompiler](https://jasperfx.github.io/lamar/documentation/compilation/). LamarCodeGeneration has a strong model for dynamically generating C# code at runtime. LamarCompiler adds runtime Roslyn support to compile assemblies on the fly and utilities to attach/create these classes. We *could* stick with `Expression` to `Lambda` compilation, but that can't really handle any kind of asynchronous code without some severe pain and it's far more difficult to reason about (Jeremy's note: I'm uniquely qualified to make this statement unfortunately).
+
+What gets dynamically generated today:
+
+* Bulk importer handling for a single entity
+* Loading entities and tracking entities in the *identity map* or version tracking
+
+What could be generated in the future:
+
+* Document metadata properties -- but sad trombone, that might have to stay with Expressions if the setters are internal/private :/
+* Much more of the `ISelector` implementations, especially since there's going to be more variability when we do the document metadata
+* Finer-grained manipulation of the `IIdentityMap` 
+
+
+## Unit Testing Approach
+
+If we introduce the runtime code generation back into Marten, that's unfortunately a non-trivial "cold start" testing issue. To soften that, I suggest we get a lot more aggressive with reusable [xUnit.Net class fixtures](https://xunit.net/docs/shared-context) between tests to reuse generated code between tests, cut way down on the sheer number of database calls by not having to frequently check the schema configuration, and other `DocumentStore` overhead. 
+
+A couple other points about this:
+
+* We need to create more unique document types so we're not having to use different configurations for the same document type. This would enable more reuse inside the testing runtime
+* Be aggressive with separate schemas for different configurations
+* We could possibly turn on xUnit.net parallel test running to speed up the test cycles
+
+
+## Document Metadata
+
+* Use the configuration and tests from [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364), but use the Lamar-backed dynamic code generation from the previous section to pull this off.
+* I think we could expand the document metadata to allow for user defined properties like "user id" or "transaction id" much the same way we'll do for the EventStore metadata. We'd need to think about how we extend the document tables and how metadata is attached to a document session
```

Comment:
I agree. We should aim to provide both built-in metadata (potentially moving there also some of the already existing columns) and user-customizable. 
About the ES metadata - for me the most important metadata are corellation id and causation id.